bundles/persistence/org.eclipse.smarthome.persistence.mapdb/src/main/java/org/eclipse/smarthome/persistence/mapdb/internal/MapDbPersistenceService.java

```diff
@@ -0,0 +1,186 @@
+@Component(service = QueryablePersistenceService.class)
+public class MapDbPersistenceService implements QueryablePersistenceService {
+
+    private static final String SERVICE_NAME = "mapdb";
+
+    private static final String DB_FOLDER_NAME = ConfigConstants.getUserDataFolder() + File.separator + "persistence" + File.separator + "mapdb";
+
+    private static final String DB_FILE_NAME = "storage.mapdb";
+
+    private final Logger logger = LoggerFactory.getLogger(MapDbPersistenceService.class);
+
+    @NonNullByDefault({})
+    private ExecutorService threadPool;
+
+    /** holds the local instance of the MapDB database */
+    @NonNullByDefault({})
+    private DB db;
+    @NonNullByDefault({})
+    private Map<String, String> map;
+
+    private transient Gson mapper = new GsonBuilder()
+            .registerTypeHierarchyAdapter(State.class, new StateTypeAdapter())
+            .create();
+
+    public void activate() {
+        logger.debug("MapDB persistence service is being activated");
+
+        threadPool = ThreadPoolManager.getPool(getClass().getSimpleName());
+
+        File folder = new File(DB_FOLDER_NAME);
+        if (!folder.exists()) {
+            if (!folder.mkdirs()) {
+                logger.warn("Failed to create one or more directories in the path '{}'", DB_FOLDER_NAME);
+                logger.warn("MapDB persistence service activation has failed.");
+                return;
+            }
+        }
+
+        File dbFile = new File(DB_FOLDER_NAME, DB_FILE_NAME);
+        db = DBMaker.newFileDB(dbFile).closeOnJvmShutdown().make();
+        map = db.createTreeMap("itemStore").makeOrGet();
+        logger.debug("MapDB persistence service is now activated");
+    }
+
+    public void deactivate() {
+        logger.debug("MapDB persistence service deactivated");
+        if (db != null) {
+            db.close();
+        }
+        threadPool.shutdown();
+    }
+
+    @Override
+    public String getId() {
+        return SERVICE_NAME;
+    }
+
+    @Override
+    public String getLabel(@Nullable Locale locale) {
+        return SERVICE_NAME;
+    }
+
+    @Override
+    public Set<PersistenceItemInfo> getItemInfo() {
+        return map.values().stream()
+                .map(this::deserialize)
+                .flatMap(MapDbPersistenceService::streamOptional)
+                .collect(Collectors.<PersistenceItemInfo>toSet());
+    }
+
+    @Override
+    public void store(Item item) {
+        store(item, item.getName());
+    }
+
+    @Override
+    public void store(Item item, String alias) {
+        if (item.getState() instanceof UnDefType) {
+            return;
+        }
+
+        logger.debug("store called for {}", alias);
+
+        State state = item.getState();
+        MapDbItem mItem = new MapDbItem();
+        mItem.setName(alias);
+        mItem.setState(state);
+        mItem.setTimestamp(new Date());
+        String json = serialize(mItem);
+        map.put(alias, json);
+        commit();
```

Comment:
In the openhab1-addons implementation there is at most 1 commit of the database every 5 seconds. I don't know the exact reasons for this but it could be an expensive operation and committing every change may also cause significant wear and tear on SD cards. Did you look into this and are there thread safety issues when there is are multiple threads in the pool?

Reply:
No I have not looked into this, the current way was suggested [above](https://github.com/eclipse/smarthome/pull/5419#discussion_r185429583), but I can see how there is a tradeoff between safety and storage wear here.

Looking at the MapDB code I don’t see any thread safety issues.